src/hotspot/share/opto/divconstants.cpp

```diff
@@ -0,0 +1,210 @@
+// x / d > x * c / m >= (x - 1) / d, which implies floor((x - 1) / d) = floor(x * c / m) since
+// there can be no integer in ((x - 1) / d, x / d)
+//
+// For -v - d + 1 <= x <= -v - 1, since v >= d - 1, we have x >= -2v
+// As a result, x * ((v + 1) / v) >= x * ((x - 2) / x) = x - 2
+//
+// x / d <= (-v - 1) / d
+// floor((x - 1) / d) = (-v - 1) / d - 1
+// x * c / m >= x * ((v + 1) / v) / d) >= (x - 2) / d
+//           >= (-v - d + 1 - 2) / d = (-v - 1) / d - 1
+// which means (-v - 1) / d >= x / d > x * c / m >= (-v - 1) / d - 1
+// This implies floor((x - 1) / d) = floor(x * c / m) for -v - 1 >= x >= -v - d + 1
+//
+// Combining all the cases gives us the conclusion.
+//
+// (3) Discussion:
+//
+// Let x be v, v - d + 1, -v, -v + d - 1, it can be seen that these bounds are indeed optimal
+//
+// (4) Implementation:
+//
+// For computation efficiency, we only consider the values m = 2**s.
+// This function does not handle the cases d being a power of 2, which means
+// that c * d is never equal to m.
+//
+// We find the value of c, m such that it satisfies the bounds for both the nonnegative
+// and negative ranges of x. This can be done by finding v_neg and v_pos and the bounds of
+// c * d - m is the intersection of (0, m / v_neg] and (0, m / v_pos). Which is (0, m / v_pos)
+// if v_pos >= v_neg and (0, m / v_neg] otherwise.
+//
+// Given v = max(v_neg, v_pos). The function inductively calculates c, rc, qv, rv such that:
+//
+// c * d - rc = 2**s with 0 < rc <= d
+// qv * v + rv = 2**s with 0 <= rv < v
+template <class T>
+void magic_divide_constants(T d, T N_neg, T N_pos, juint min_s, T& c, bool& c_ovf, juint& s) {
+  static_assert(std::is_unsigned<T>::value, "calculations must be done in the unsigned domain");
+  assert(!is_power_of_2(d), "this case should be handled separately");
+  assert(d <= N_neg || d <= N_pos, "this should just be idealized to 0");
+  constexpr T min_signed = std::numeric_limits<std::make_signed_t<T>>::min();
+  T v_neg = N_neg < d - 1 ? 0 : N_neg - ((N_neg - d + 1) % d);
+  T v_pos = N_pos < d - 1 ? 0 : N_pos - ((N_pos - d + 1) % d);
+  T v = MAX2(v_neg, v_pos);
+  bool tolerate_equal = v_neg > v_pos;
+
+  // base case, s = 0
+  s = 0;
+  c_ovf = false;
+  c = 1;
+  T rc = d - 1;
+  bool qv_ovf = false;
+  T qv = 0;
+  T rv = 1;
+
+  while (true) {
+    if (s >= min_s) {
+      if (qv_ovf || rc < qv || (rc == qv && (tolerate_equal || rv > 0))) {
+        break;
+      }
+    }
+    assert(!c_ovf, "must be");
+    s++;
+
+    T new_rc = rc * 2;
+    if (new_rc < rc || new_rc > d) {
+      c_ovf = c > min_signed;
+      c = c * 2 - 1;
+      rc = new_rc - d;
+    } else {
+      c_ovf = c >= min_signed;
+      c = c * 2;
+      rc = new_rc;
+    }
+
+    T new_rv = rv * 2;
+    if (new_rv < rv || new_rv >= v) {
+      qv_ovf = qv >= min_signed;
+      qv = qv * 2 + 1;
+      rv = new_rv - v;
+    } else {
+      qv_ovf = qv >= min_signed;
+      qv = qv * 2;
+      rv = new_rv;
+    }
```

Comment:
One could perhaps avoid overflows in computing `rc` and `rv`, and simplify the corresponding tests, like so (not sure if this improves anything in practical terms, though):
```suggestion
    if (d - rc < rc) {  // 2 * rc > d
      c_ovf = c > min_signed;
      c += c - 1;
      rc -= d - rc;  // rc = 2 * rc - d
    } else {  // 2 * rc <= d
      c_ovf = c >= min_signed;
      c += c;
      rc += rc;  // rc = 2 * rc
    }

    if (rv >= v - rv) {  // 2 * rv >= v
      qv_ovf = qv >= min_signed;
      qv += qv + 1;
      rv -= v - rv;  // rv = 2 * rv - v
    } else {  // 2 * rv < v
      qv_ovf = qv >= min_signed;
      qv += qv;
      rv += rv;  // rv = 2 * rv
    }
```

Reply:
That's a good idea, I have changed the formulae as your suggestion.

Thanks a lot for your effort, this is really dense in mathematical transformations, I really appreciate it.